src/vos/ilog.c

```diff
@@ -0,0 +1,1474 @@
+
+/** Register a set of custom callbacks for tx_id */
+void
+ilog_set_callbacks(const struct ilog_callbacks *cbs)
+{
+	ilog_cbs = *cbs;
+}
+
+/** The ilog is split into two parts.   If there is one entry, the ilog
+ *  is embedded into the root df struct.   If not, a b+tree is used.
+ *  The tree is used more like a set where only the key is used.
+ */
+
+struct ilog_tree {
+	umem_off_t	it_root;
+	uint64_t	it_embedded;
+};
+
+struct ilog_root {
+	union {
+		struct ilog_id		lr_id;
+		struct ilog_tree	lr_tree;
+	};
+	bool				lr_punch;
+	uint32_t			lr_magic;
+};
+
+struct ilog_context {
+	/** Root pointer */
+	struct ilog_root		*ic_root;
+	/** umem offset of root pointer */
+	umem_off_t			 ic_root_off;
+	/** umem instance */
+	struct umem_instance		 ic_umm;
+	/** ref count for iterator */
+	uint32_t			 ic_ref;
+	/** In pmdk transaction marker */
+	bool				 ic_in_txn;
+};
+
+#define DF_ID		"epoch:"DF_U64" tx_id:0x"DF_X64
+#define DP_ID(id)	(id).id_epoch, (id).id_tx_id
+#define DF_VAL		"punch:%s"
+#define DP_VAL(punch)	punch ? " true" : "false"
+
+struct prec {
+	bool	p_punch;
+	int32_t	p_magic;
+};
+
+static inline struct prec *
+rec2prec(struct btr_record *rec)
+{
+	return (struct prec *)&rec->rec_off;
+}
+
+static inline bool *
+rec2punch(struct btr_record *rec)
+{
+	return &rec2prec(rec)->p_punch;
+}
+
+D_CASSERT(sizeof(struct ilog_id) == sizeof(struct ilog_tree));
+D_CASSERT(sizeof(struct ilog_root) == sizeof(struct ilog_df));
+/** We hijack the value offset to store the actual value inline */
+D_CASSERT(sizeof(struct prec) <= sizeof(((struct btr_record *)0)->rec_off));
+
+/**
+ * Customized functions for btree.
+ */
+
+/** size of hashed-key */
+static int
+ilog_hkey_size(void)
+{
+	return sizeof(struct ilog_id);
+}
+
+static int
+ilog_rec_msize(int alloc_overhead)
+{
+	/** No extra allocation for ilog entries */
+	return 0;
+}
+
+/** generate hkey */
+static void
+ilog_hkey_gen(struct btr_instance *tins, d_iov_t *key_iov, void *hkey)
+{
+	D_ASSERT(key_iov->iov_len == sizeof(struct ilog_id));
+	memcpy(hkey, key_iov->iov_buf, sizeof(struct ilog_id));
+}
+
+/** compare the hashed key */
+static int
+ilog_hkey_cmp(struct btr_instance *tins, struct btr_record *rec, void *hkey)
+{
+	struct ilog_id	*k1 = (struct ilog_id *)&rec->rec_hkey[0];
+	struct ilog_id	*k2 = (struct ilog_id *)hkey;
+
+	if (k1->id_epoch < k2->id_epoch)
+		return BTR_CMP_LT;
+
+	if (k1->id_epoch > k2->id_epoch)
+		return BTR_CMP_GT;
+
+	return BTR_CMP_EQ;
+}
+
+/** create a new key-record, or install an externally allocated key-record */
+static int
+ilog_rec_alloc(struct btr_instance *tins, d_iov_t *key_iov,
+	      d_iov_t *val_iov, struct btr_record *rec)
+{
+	struct prec *prec = rec2prec(rec);
+
+	D_ASSERT(val_iov->iov_len == sizeof(bool));
+	/** Note the D_CASSERT above ensures that rec_off is large enough
+	 *  to fit the value without allocating new memory.
+	 */
+	prec->p_punch = *(bool *)val_iov->iov_buf;
+	prec->p_magic = 1; /* So free is always called */
```

Comment:
I'm using the rec_off field to store the data rather than allocating extra memory because the data we store is just the punch value.

The magic is needed because common btree code doesn't call rec_free if rec_off is UMOFF_NULL, which it would be for a non-punch creation timestamp.